include/knowhere/index/index_node.h

```diff
@@ -95,41 +94,65 @@ class IndexNode : public Object {
+        // The `range_search_k` is used to early terminate the iterator-search.
+        //  When the number of range-valid results exceeds range_search_k, the `further_bound` will be updated.
+        //  "-1" means no early termination, `further_bound` will not be updated.
+        // Note that the number of results is not guaranteed to be exactly range_search_k, it may be more or less.
+        float further_bound = base_cfg.radius.value();
+        const int32_t range_search_k = base_cfg.range_search_k.value();
+        LOG_KNOWHERE_DEBUG_ << "range_search_k: " << range_search_k;
+
+        const bool the_larger_the_closer = IsMetricType(base_cfg.metric_type.value(), metric::IP) ||
+                                           IsMetricType(base_cfg.metric_type.value(), metric::COSINE) ||
+                                           IsMetricType(base_cfg.metric_type.value(), metric::BM25);
+        auto is_first_closer = [&the_larger_the_closer](const float dist_1, const float dist_2) {
+            return the_larger_the_closer ? dist_1 > dist_2 : dist_1 < dist_2;
+        };
+        auto too_close = [&is_first_closer, &closer_bound](float dist) { return is_first_closer(dist, closer_bound); };
+        auto too_far = [&is_first_closer, &further_bound](float dist) { return !is_first_closer(dist, further_bound); };
 
         const auto its = its_or.value();
         const auto nq = its.size();
-
         std::vector<std::vector<int64_t>> result_id_array(nq);
         std::vector<std::vector<float>> result_dist_array(nq);
-        const bool similarity_metric = IsMetricType(base_cfg.metric_type.value(), metric::IP) ||
-                                       IsMetricType(base_cfg.metric_type.value(), metric::COSINE);
-        const bool has_range_filter = range_filter != defaultRangeFilter;
-        constexpr size_t k_min_num_consecutive_over_radius = 16;
-        const auto range_search_level = base_cfg.range_search_level.value();
+
+        constexpr size_t k_min_num_consecutive_over_further_bound = 16;
+        const auto range_search_level = base_cfg.range_search_level.value();  // from 0 to 0.5
         LOG_KNOWHERE_DEBUG_ << "range_search_level: " << range_search_level;
+
         auto task = [&](size_t idx) {
+            // max-heap, use top as the further_bound if size == range_search_k
+            std::priority_queue<float, std::vector<float>, decltype(is_first_closer)> early_stop_further_bounds(
+                is_first_closer);
             auto it = its[idx];
             size_t num_next = 0;
-            size_t num_consecutive_over_radius = 0;
+            size_t num_consecutive_over_further_bound = 0;
             while (it->HasNext()) {
                 auto [id, dist] = it->Next();
                 num_next++;
-                // too close
-                if (has_range_filter && (similarity_metric ? dist > range_filter : dist < range_filter)) {
+                if (has_closer_bound && too_close(dist)) {
```

Comment:
maybe capture `has_closer_bound` in `too_close` as well

Reply:
maybe it is better to separate.